unittests/Interpreter/InterpreterTests.cpp

```diff
@@ -0,0 +1,442 @@
+    return x;
+  }
+
+  KnownBits clearLowest(KnownBits x) {
+    for (int i = 0; i < x.getBitWidth(); i++) {
+      if (!x.Zero[i] && !x.One[i]) {
+	x.Zero.setBit(i);
+	return x;
+      }
+    }
+    assert(false);
+    return x;
+  }
+
+  KnownBits bruteForce(KnownBits x, KnownBits y, Inst::Kind Pred) {
+    if (!isConcrete(x))
+      return merge(bruteForce(setLowest(x), y, Pred),
+		   bruteForce(clearLowest(x), y, Pred));
+    if (!isConcrete(y))
+      return merge(bruteForce(x, setLowest(y), Pred),
+		   bruteForce(x, clearLowest(y), Pred));
+    auto xc = x.getConstant();
+    auto yc = y.getConstant();
+
+    KnownBits res(x.getBitWidth());
+    switch (Pred) {
+    case Inst::AddNUW:
+    case Inst::AddNSW:
+    case Inst::AddNW:
+    case Inst::Add:
+    {
+      auto rc = xc + yc;
+      res.One = rc;
+      res.Zero = ~rc;
+    }
+    break;
+    case Inst::SubNUW:
+    case Inst::SubNSW:
+    case Inst::SubNW:
+    case Inst::Sub:
+    {
+      auto rc = xc - yc;
+      res.One = rc;
+      res.Zero = ~rc;
+    }
+    break;
+    case Inst::Mul:
+    {
+      auto rc = xc * yc;
+      res.One = rc;
+      res.Zero = ~rc;
+    }
+      break;
+    case Inst::UDiv:
+    {
+      if (yc != 0) {
+	auto rc = xc.udiv(yc);
+	res.One = rc;
+	res.Zero = ~rc;
+      }
+    }
+    break;
+    case Inst::URem:
+    {
+      if (yc != 0) {
+	auto rc = xc.urem(yc);
+	res.One = rc;
+	res.Zero = ~rc;
+      }
+    }
+    break;
+    case Inst::And:
+    {
+      auto rc = xc & yc;
+      res.One = rc;
+      res.Zero = ~rc;
+    }
+    break;
+    case Inst::Or:
+    {
+      auto rc = xc | yc;
+      res.One = rc;
+      res.Zero = ~rc;
+    }
+    break;
+    case Inst::Xor:
+    {
+      auto rc = xc ^ yc;
+      res.One = rc;
+      res.Zero = ~rc;
+    }
+    break;
+    case Inst::Shl:
+    {
+      auto rc = xc << yc;
+      res.One = rc;
+      res.Zero = ~rc;
+    }
+    break;
+    case Inst::LShr:
+    {
+      if (yc.getLimitedValue() < res.getBitWidth()) {
+	auto rc = xc.lshr(yc.getLimitedValue());
+	res.One = rc;
+	res.Zero = ~rc;
+      }
+    }
+    break;
+    case Inst::AShr:
+    {
+      if (yc.getLimitedValue() < res.getBitWidth()) {
+	auto rc = xc.ashr(yc.getLimitedValue());
+	res.One = rc;
+	res.Zero = ~rc;
+      }
+    }
+    break;
+    case Inst::Eq:
+    {
+      res = KnownBits(1);
+      if (xc.eq(yc))
+	res.One.setBit(0);
+      else
+      	res.Zero.setBit(0);
+    }
+    break;
+    case Inst::Ne:
+    {
+      res = KnownBits(1);
+      if (xc.ne(yc))
+	res.One.setBit(0);
+      else
+	res.Zero.setBit(0);
+    }
+    break;
+    case Inst::Ult:
+    {
+      res = KnownBits(1);
+      if (xc.ult(yc))
+	res.One.setBit(0);
+      else
+	res.Zero.setBit(0);
+    }
+    break;
+    case Inst::Slt:
+    {
+      res = KnownBits(1);
+      if (xc.slt(yc))
+	res.One.setBit(0);
+      else
+	res.Zero.setBit(0);
+    }
+    break;
+    case Inst::Ule:
+    {
+      res = KnownBits(1);
+      if (xc.ule(yc))
```

Comment:
can you just return the result of the ule call? why do you need to make a new object of the knownbits type?

Reply:
result of the ule call is an APInt and KnownBits doens't have a public constructor that takes APInt.